gsplat/strategy.py

```diff
@@ -0,0 +1,552 @@
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Any, Callable, DefaultDict, Dict, List, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+from gsplat import quat_scale_to_covar_preci
+from gsplat.relocation import compute_relocation
+from gsplat.utils import normalized_quat_to_rotmat
+
+
+class Strategy(ABC):
+    """Base class for the GS optimization strategy."""
+
+    def __init__(
+        self,
+        params: torch.nn.ParameterDict,
+        optimizers: List[torch.optim.Optimizer],
```

Comment:
wouldn't this need to be a dict matching the key in the paramdict to the optimizer?

Reply:
A single optimizer is enough as long as the param group has a "name" field defined. The reason I'm putting a List here is in case user wants to use different optimizer for different field. E.g., SGD for color and Adam for the rest

https://github.com/nerfstudio-project/gsplat/blob/main/examples/simple_trainer.py#L217-L224

Reply:
I see, so it will automatically loop through all optimizers and check if any of the param group names match the paramdict?

Reply:
in nerfstudio there is a separate `torch.nn.Optimizer` for each param group, so we should check if passing in a list of those works

Reply:
List of optimizers also works as long as the param group has a "name" field.

The logic is that it will loop over the list of optimizers and every parameter group in it to find a match, based on the "name" field

Reply:
Actually, maybe its more user friendly to just use a dict optimizer and force each parameter has a seperate optimizer?